java-checks/src/main/java/org/sonar/java/checks/spring/NonSingletonAutowiredInSingletonCheck.java

```diff
@@ -0,0 +1,208 @@
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S6832")
+public class NonSingletonAutowiredInSingletonCheck extends IssuableSubscriptionVisitor {
+  private static final String SCOPED_ANNOTATION = "org.springframework.context.annotation.Scope";
+  private static final String AUTOWIRED_ANNOTATION = "org.springframework.beans.factory.annotation.Autowired";
+  private static final String JAVAX_INJECT_ANNOTATION = "javax.inject.Inject";
+  private static final String JAKARTA_INJECT_ANNOTATION = "jakarta.inject.Inject";
+  private static final Set<String> AUTO_WIRING_ANNOTATIONS = Set.of(AUTOWIRED_ANNOTATION, JAVAX_INJECT_ANNOTATION, JAKARTA_INJECT_ANNOTATION);
+  private static final Set<String> SINGLETON_LITERALS = Set.of("singleton", "\"singleton\"");
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.ANNOTATION, Tree.Kind.CONSTRUCTOR);
+  }
+
+  /**
+   * This rule reports an issue when a Singleton Bean auto-wires a non-Singleton Bean via constructor, field or method parameter.
+   */
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.is(Tree.Kind.ANNOTATION)) {
+      analyzeAnnotation((AnnotationTree) tree);
+    }
+
+    if (tree.is(Tree.Kind.CONSTRUCTOR)) {
+      analyzeSingleArgumentConstructor((MethodTree) tree);
+    }
+  }
+
+  private void analyzeAnnotation(AnnotationTree annotationTree) {
+    if (!isAutoWiringAnnotation(annotationTree)) {
+      return;
+    }
+
+    var annotatedSymbol = Optional.ofNullable(annotationTree.parent()).map(Tree::parent).orElse(null);
+    if (annotatedSymbol == null) {
+      return;
+    }
+
+    if (annotatedSymbol.is(Tree.Kind.VARIABLE)) {
+      analyzeAnnotatedFieldOrParameter((VariableTree) annotatedSymbol);
+
+    } else if (annotatedSymbol.is(Tree.Kind.METHOD)) {
+      analyzeAnnotatedSetter((MethodTree) annotatedSymbol);
+
+    } else if (annotatedSymbol.is(Tree.Kind.CONSTRUCTOR)) {
+      analyzeAnnotatedConstructor((MethodTree) annotatedSymbol);
+    }
+  }
+
+  private void analyzeAnnotatedFieldOrParameter(VariableTree annotatedVar) {
+    if (isClassField(annotatedVar)) {
+      getEnclosingClass(annotatedVar.symbol().enclosingClass())
+        .ifPresent(enclosingClassTree -> reportIfNonSingletonInSingleton(enclosingClassTree, annotatedVar, "autowired field"));
+    }
+
+    if (isSetterParameter(annotatedVar) || isConstructorParameter(annotatedVar)) {
+      getEnclosingClass(annotatedVar.symbol().enclosingClass())
+        .ifPresent(enclosingClassTree -> reportIfNonSingletonInSingleton(enclosingClassTree, annotatedVar, "autowired parameter"));
```

Comment:
The only difference seems to be the message to me - can we save a couple of lines of code by only setting a different message in the if/else getting the class + reporting only once?